rpc/src/account_history.rs

```diff
@@ -0,0 +1,226 @@
+use {
+    crossbeam_channel::{Receiver, RecvTimeoutError},
+    csv::{ReaderBuilder, Trim},
+    solana_measure::measure::Measure,
+    solana_metrics::datapoint_info,
+    solana_runtime::bank::Bank,
+    solana_sdk::{account::Account, clock::Slot, pubkey::Pubkey},
+    std::{
+        collections::{BTreeMap, HashMap, HashSet},
+        path::Path,
+        str::FromStr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc, RwLock, RwLockWriteGuard,
+        },
+        thread::{self, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+#[derive(Debug, Clone)]
+pub struct AccountHistoryConfig {
+    pub num_slots: usize,
+    pub updates_only: bool,
+    pub account_keys: AccountKeys,
+}
+
+pub type AccountHistory = BTreeMap<Slot, HashMap<Pubkey, Account>>;
+pub type AccountKeys = HashSet<Pubkey>;
+
+pub struct RpcAccountHistoryService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl RpcAccountHistoryService {
+    pub fn new(
+        config: &AccountHistoryConfig,
+        account_keys: Arc<RwLock<AccountKeys>>,
+        account_history: Arc<RwLock<AccountHistory>>,
+        account_history_receiver: Receiver<Arc<Bank>>,
+        exit: &Arc<AtomicBool>,
+    ) -> Self {
+        let exit = exit.clone();
+        let config = config.clone();
+        let thread_hdl = Builder::new()
+            .name("solana-account-history".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+                if let Err(RecvTimeoutError::Disconnected) = Self::receive_bank(
+                    &config,
+                    &account_keys,
+                    &account_history,
+                    &account_history_receiver,
+                ) {
+                    break;
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    fn receive_bank(
+        config: &AccountHistoryConfig,
+        account_keys: &Arc<RwLock<AccountKeys>>,
+        account_history: &Arc<RwLock<AccountHistory>>,
+        account_history_receiver: &Receiver<Arc<Bank>>,
+    ) -> Result<(), RecvTimeoutError> {
+        let frozen_bank = account_history_receiver.recv_timeout(Duration::from_secs(1))?;
+        debug!(
+            "rpc account-history service received bank: {:?}",
+            frozen_bank.slot()
+        );
+
+        let r_account_keys = account_keys.read().unwrap();
+        let mut measure_collect = Measure::start("collect-account-history");
+        let slot_accounts =
+            Self::collect_accounts(&frozen_bank, &r_account_keys, config.updates_only);
+        measure_collect.stop();
+        drop(r_account_keys);
+
+        let mut measure_write = Measure::start("write-account-history");
+        let mut w_account_history = account_history.write().unwrap();
+        w_account_history.insert(frozen_bank.slot(), slot_accounts);
```

Comment:
Assuming this will eventually be persisted to something like the blockstore?

Reply:
I don't think that's needed at the moment. Since this service is intended to replace account subscriptions for the motivating use-case, I think it's expected that updates that occur when the node is offline will be unavailable. But yeah, blockstore backup is def an option if we need stronger persistence.